Implemented standard tensors for elasticity.

jppelteret opened the pull request:
Refers to discussion in dealii/code-gallery#7 .

TODO:
- [x] Implement deformation gradient and strain tensors
- [x] Standard tensor derivatives?
- [x] Check doxygen documentation (can't build doxygen at the moment)
- [x] Update step-44

davydden commented on source/base/standard_tensors.inst.in:
```diff
@@ -0,0 +1,27 @@
+    class StandardTensors<deal_II_dimension,number>;
+}
+
+//for (deal_II_dimension : DIMENSIONS; number : COMPLEX_SCALARS)
```
any problems here?

jppelteret replied:
Didn't compile :-( I'll just remove this, because I don't see application for it at the moment.

davydden commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,208 @@
+};
+
+
+
```
you may need to add `#ifndef DOXYGEN` here

davydden commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,208 @@
+  return Dev_P_T;
+}
+
+
```
and `#endif // DOXYGEN` here.

davydden commented:
nice :+1: 🎉

davydden commented:
suggested TODO: update step-44 to use those and mention them in introduction.

bangerth commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,208 @@
+   * $\mathbf{I} \cdot \{ \bullet \} = \{ \bullet \} \cdot \mathbf{I} =
+   * \{ \bullet \}$ and
+   * $\mathbf{I} : \{ \bullet \} = \textrm{trace} \{ \bullet \}$.
+   * This aligns with the second-order SymmetricTensor returned by unit_symmetric_tensor.
```
Can you explain here what concrete product `\cdot` and `:` correspond to?

bangerth replied:
add `()` to the function name at the end of the line so doxygen cross references it

bangerth commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,208 @@
+   *
+   * This is defined such that, for any second-order SymmetricTensor,
+   * the following holds:
+   * $\mathbf{II} : \{ \bullet \} = \{ \bullet \} : \mathbf{II} =
```
Same here with the product

bangerth commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,208 @@
+  static const SymmetricTensor<2, dim, Number> I;
+
+  /**
+   * The fourth order referential/spatial unit SymmetricTensor $\mathcal{I}$.
```
Should this be `II` in the formula?

bangerth commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,208 @@
+   * [1/\textrm{dim}][ \{ \bullet\} :\mathbf{I}]\mathbf{I}$
+   * and, therefore,
+   * $\textrm{trace} \textrm{dev} \{ \bullet \} = 0$.
+   * This aligns with the fourth-order SymmetricTensor returned by deviator_tensor.
```
`()`

bangerth commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,208 @@
+   * $\textrm{Dev} \{ \bullet \} = J^{-2/\textrm{dim}}[ \{ \bullet \} -
+   * [1/\textrm{dim}][\mathbf{C} : \{ \bullet\}] \mathbf{C}^{-1} ]$
+   * with
+   * $\mathbf{C} = \mathbf{F}^{T}*\mathbf{F}$ and $J = \textrm{det}\mathbf{F}$
```
What is the star operator here?

bangerth commented:
Nice.

Out of curiosity: the tensors you define all have real, scalar elements. That's independent of the fact that they may be multiplied by tensors that have other underlying scalars. Why not always create objects whose scalar type is simply `double`? I don't really see a good reason why anyone would create a `StandardTensors<2,std::complex<double>>`.

jppelteret commented:
Thanks @davydden  and @bangerth for the comments. Sorry about the doxygen issues - I couldn't install doxygen last night to check for errors. Fixes to come.

Regarding the number type, I want to be able to use these tensors with auto-differentiable number types. It doesn't really make a difference for most of the tensors listed here so far, but it does matter for `Dev_P`. I could templatize this function alone (instead of the class), but then one will have to do similar when extending the class to include more field-dependent functions. Thoughts?

Lastly, a point of debate/discussion:  Since this was taken directly from `step-44`, I kept the `StandardTensors` class name. Thinking towards the future, I was wondering if I should change this to a namespace and instead have the class `StandardTensors::Elasticity`. This is more descriptive and if we are to add other standard tensors for different physics problems then it would be more easily extended. Also, if the literature clashes on tensor names, then this potential issue or difficulty could be partially mitigated. Are you in favour of this change?

davydden commented:
> StandardTensors::Elasticity

:+1:

bangerth commented:
I think it would be better to templatize only those functions that really need it. I understand that you may want to *multiply* these tensors by tensors that have AD scalars, but that should already be possible. I don't suspect that you will ever want to take derivatives *with regard to the elements of the standard tensors*.

 You may even be able to then make what's currently a class into a namespace. (Whether or not that is nested in another namespace or not, on which I don't really have an opinion, though names such as `Dev_P` are clearly elasticity specific.)

jppelteret commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,513 @@
+StandardTensors::Elasticity<dim>::E (const Tensor<2, dim, Number> &F)
+{
+ // See Wriggers (2008) p 23 equ. 3.15
+ return Number(0.5)*(C(F) - unit_symmetric_tensor<dim,Number>());
```
@bangerth This is an example of a function that needs `I == unit_symmetric_tensor<dim>()` to have a `number` type. Without it, one gets this error
```bash
/Users/jp/Programs/GitHub/deal.II/dealii/include/deal.II/base/standard_tensors.h:390:27: error: invalid operands to binary expression ('SymmetricTensor<2, 1, float>' and 'const SymmetricTensor<2, 1>')
 return Number(0.5)*(C(F) - I);
```
presumably because for the float instantiation the compiler's having trouble casting up the return type of `C(F)`  to a `SymmetricTensor<2,dim,double>` and then casting it down again to a `SymmetricTensor<2,dim,float>`.

I agree that in theory defining `I` without a `number` type is sensible, but  when using it for additive operations with mixed number types it becomes a pain to use.

bangerth replied:
Would it work if we made sure that `operator-` and `operator+` do what they are expected to do? I would expect that we could declare them as something like
```
  template <int dim, typename Number1, typename Number2>
  SymmetricTensor<2,dim,decltype(Number1() - Number2())>
  operator- (....) {....}
```
This requires C++11, but would at least do the right thing.

I'm just wondering where the pain is best spent. Yes, we could write things in such a way that we track the template argument through everything. Or we could spend the time overloading a few operators. I don't know what's better, but I've learned that having template arguments on everything is often no fun either.

jppelteret commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,513 @@
+  // See Wriggers p46 equ 3.125
+  SymmetricTensor<4,dim,Number> Dev_P = outer_product(C_inv,C);  // Dev_P = C_inv_x_C
+  Dev_P /= -dim;                                                 // Dev_P = -[1/dim]C_inv_x_C
+  Dev_P += SymmetricTensor<4,dim,Number>(II);                    // Dev_P = II - [1/dim]C_inv_x_C
```
@bangerth Here's another example where I've had to cast a standard tensor type.

bangerth replied:
Same here -- as long as `Number() + double()` makes sense, it should make sense to add `SymmetricTensor<4,dim,double>` to `Dev_P`.

bangerth commented on include/deal.II/base/standard_tensors.h:
```diff
@@ -0,0 +1,513 @@
+   */
+  template <typename Number>
+  static SymmetricTensor<2, dim, Number>
+  C (const Tensor<2, dim, Number> &F);
```
In the example below where you use `C`, it struck me that for it to make sense, `C(F)` necessarily had to have no units. I think that's actually true here. But I think it would make sense for each of these functions to state what the units of all quantities are. (Here, of `F` and of what the function returns.) We've found in ASPECT that diligently documenting all units of everything is a really useful process in finding subtle bugs.

maieneuro commented on include/deal.II/physics/elasticity/kinematics.h:
```diff
@@ -0,0 +1,278 @@
+   const Tensor<2,spacedim> & dF_dt)
+{
+  // See Holtzapfel (2007) eq 2.149
+  return l(F,dF_dt) - dF,dF_dt) ;
```
missing opening bracket in `d(F,dF_dt)` ?

jppelteret replied:
Thanks @maieneuro ! I haven't yet finished implementing tests so didn't pick it up :-) I'll fix this shortly.

jppelteret replied:
@maieneuro Whilst you're looking around, do think that there's any relevant functionality that I've missed?

maieneuro replied:
Nothing that I can think of, this will be very handy.

maieneuro commented on include/deal.II/physics/elasticity/kinematics.h:
```diff
@@ -0,0 +1,278 @@
+Physics::Elasticity::Kinematics<dim,Number>::F (const Tensor<2, dim, Number> &Grad_u)
+{
+// See Wriggers (2008) p 23 equ. 3.14
+  return Physics::Elasticity::StandardTensors<dim>::I + Grad_u;
```
I vaguely remember this biting me in the past so I usually do the equivalent of `static_cast<Tensor<...>>(I) + Grad_u`. It may have been a red-herring from me trying to fix something else, but it's worth a double check.

jppelteret replied:
I also used to explicitly cast all SymmetricTensors to standard tensors (e.g. [here](http://www.dealii.org/developer/doxygen/deal.II/step_44.html#Quadraturepointhistory)) but this seems to no longer be necessary. I also can't remember why we needed to do this in the first place... However, I will double check it as suggested.

jppelteret replied:
Agreed. I’ll do this.

jppelteret commented:
Agreed. I’ll do this.
> On 15 Nov 2016, at 13:22, Denis Davydov <notifications@github.com> wrote:
> 
> @davydden commented on this pull request.
> 
> In include/deal.II/physics/elasticity/kinematics.h <https://github.com/dealii/dealii/pull/3034>:
> 
> > +  }
> +}
> +
> +
> +
> +// ------------------------- inline functions ------------------------
> +
> +
> +
> +template <int dim, typename Number>
> +inline
> +Tensor<2, dim, Number>
> +Physics::Elasticity::Kinematics<dim,Number>::F (const Tensor<2, dim, Number> &Grad_u)
> +{
> +// See Wriggers (2008) p 23 equ. 3.14
> +  return Physics::Elasticity::StandardTensors<dim>::I + Grad_u;
> i guess adding a unit test for this part would then make a lot of sense.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/dealii/dealii/pull/3034>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AJ-dwlM6eQJJPkXjG6kz5clsBZOn6hh-ks5q-aQIgaJpZM4Jv9ip>.
>

jppelteret commented:
Ok, I think that I've done enough to justify asking for some further feedback on this. It would be great to tick this one off my list sooner rather than later.

@bangerth, I think that I attended to all of your initial comments. Thanks again for the input!
@davydden I chose to stick with `Elasticity::StandardTensors` because then it doesn't break the neatly nested structure that already arranged 😬

The new set of namespaces / classes has the following hierarchy:
- NS Physics
  - NS Transformations
    - NS Contravariant
    - NS Covariant
    - NS Piola
  - NS Elasticity
    - C StandardTensors
    - NS Kinematics
    - [NS Kinetics] (possible later extension to compute stresses etc.; maybe a job for the weekend)

I've produced a couple of tests based on `step-44` that show that (a large portion of) the implemented functions and definitions work as expected. For the remaining functions, I'll have to think a little more carefully as to how one can construct small test cases for them.

Once this PR has reached the stage of general approval, I'll update `step-44` itself (there are only a couple of lines to change) and rebase etc.

davydden commented on include/deal.II/base/tensor.h:
```diff
@@ -1921,6 +1921,36 @@ transpose (const Tensor<2,dim,Number> &t)
 
 
 /**
+ * Return the adjoint of the given tensor.
```
i would add formula here so that it's clearer for users.

jppelteret replied:
I've done this - Github hasn't registered the change on the line below.

davydden commented on include/deal.II/base/tensor.h:
```diff
@@ -1921,6 +1921,36 @@ transpose (const Tensor<2,dim,Number> &t)
+
+
+/**
+ * Return the cofactor of the given tensor.
```
same here.

jppelteret replied:
Ditto

davydden commented on include/deal.II/physics/elasticity/kinematics.h:
```diff
@@ -0,0 +1,365 @@
+       * The differential operator $\nabla_{0}$ is defined as
+       * $\frac{\partial}{\partial \mathbf{X}}$.
+       *
+       * @param[in] Grad_u Material displacement gradient tensor $\nabla_{0}\mathbf{u}$
```
i guess you don't need those, as you explained it in the text already.

davydden commented on include/deal.II/physics/elasticity/kinematics.h:
```diff
@@ -0,0 +1,365 @@
+
+      /**
+       * Returns the isochoric counterpart of the deformation gradient Tensor,
+       * deformation gradient Tensor @p F itself.
```
here and in the next one, i am puzzled with 

> ...the deformation gradient Tensor, deformation gradient Tensor @p F itself.

maybe it's just a long day...

davydden commented on include/deal.II/physics/elasticity/standard_tensors.h:
```diff
@@ -0,0 +1,355 @@
+       * Returns the derivative of the volumetric Jacobian
+       * $J = \text{det} \mathbf{F}$ with respect to the right Cauchy-Green
+       * tensor, as constructed from the deformation gradient tensor @p F.
+       */
```
add formula here?

davydden commented on include/deal.II/physics/physics.h:
```diff
@@ -0,0 +1,118 @@
+   * <h3>Notation</h3>
+   *
+   * References for this notation include:
+   *    - G.A. Holzapfel (2001),
```
you can put it in a bibtex form (I don't recall exactly the syntax, for sure used in FE_Enriched)

davydden commented on include/deal.II/physics/physics.h:
```diff
@@ -0,0 +1,118 @@
+   * tensors (matrices) onto themselves in much the same way as matrices map
+   * vectors onto vectors.
+   * To provide some context to the implemented class members and functions,
+   * consider the following fundamental operations performed tensors with special
```
performed ON ?

davydden commented on include/deal.II/physics/physics.h:
```diff
@@ -0,0 +1,118 @@
+   * The fourth-order SymmetricTensor returned by identity_tensor() is
+   * $\mathcal{S}$, and that returned by deviator_tensor() is $\mathcal{P}$.
+   *
+   * For convenience we will predefine come commonly referenced tensors and
```
come -> some

davydden commented on source/physics/elasticity/kinematics.inst.in:
```diff
@@ -0,0 +1,68 @@
+// ---------------------------------------------------------------------
+
+
+for (deal_II_dimension : DIMENSIONS; number : REAL_SCALARS)
```
does something not work for complex-valued?

davydden replied:
probably not. I am ok to keep it as is.

davydden commented:
Agreed

> 3 дек. 2016 г., в 12:28, Jean-Paul Pelteret <notifications@github.com> написал(а):
> 
> @jppelteret commented on this pull request.
> 
> In source/physics/elasticity/kinematics.inst.in:
> 
> > +//
> +// Copyright (C) 2016 by the deal.II authors
> +//
> +// This file is part of the deal.II library.
> +//
> +// The deal.II library is free software; you can use it, redistribute
> +// it, and/or modify it under the terms of the GNU Lesser General
> +// Public License as published by the Free Software Foundation; either
> +// version 2.1 of the License, or (at your option) any later version.
> +// The full text of the license can be found in the file LICENSE at
> +// the top level of the deal.II distribution.
> +//
> +// ---------------------------------------------------------------------
> +
> +
> +for (deal_II_dimension : DIMENSIONS; number : REAL_SCALARS)
> I don't think that it makes sense to define complex kinematic variables. Its there a scenario in which one has "imaginary" deformation?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

davydden commented on doc/news/changes.h:
```diff
@@ -223,6 +223,28 @@ inconvenience this causes.
 
 <ol>
 
+ <li> New: Implemented within the Physics::Elasticity namespace is a class
```
> Implemented within.. is a class ... that...

something is wrong here, i guess you can just say `Implement several standard tensors ... within Physics::Elasticity::StandardTensors and Physics::Elasticity::Kinematics` or alike.

davydden commented on doc/news/changes.h:
```diff
@@ -223,6 +223,28 @@ inconvenience this causes.
+ (Jean-Paul Pelteret, Andrew McBride, 2016/12/05)
+ </li>
+
+ <li> New: There is now a namespace dedicated to defining useful functions
```
you can say what namespace here so that it is hyperlinked later.

davydden commented on examples/step-44/step-44.cc:
```diff
@@ -431,44 +441,6 @@ namespace Step44
-
-// Now we define some frequently used second and fourth-order tensors:
-  template <int dim>
-  class StandardTensors
```
nice :smile:

davydden commented on include/deal.II/physics/elasticity/standard_tensors.h:
```diff
@@ -0,0 +1,360 @@
+
+      /**
+       * The fourth-order spatial deviatoric tensor. Also known as the deviatoric
+       * operator, this tensor projects a second-order SymmetricTensor onto a
```
i think it projects arbitrary second order tensor onto the same second order tensor with trace being zero. In other words, i would say `dev_P : A = A - 1/3tr(A) I`, which is exactly what you wright below under the first `=` sign.

davydden replied:
>  If A is not symmetric then this will return sym(A) - 1/dim tr(A) I.

i don't think so:

`[dev_P : A]_{ij} = [\delta_{ik} \delta_{jl} -...] A_{kl} = A_{ij} - ...` ?

davydden replied:
ok, i think i know what's the problem: I am using a general fourth order unit tensor, whereas you refer to a symmetric variant, i.e. : `0.5[\delta_{ik}{jl}+\delta_{il}{jk}`.

davydden replied:
... which is essentially an average of the `I` such that `I:A=A` and `It` such that `It:A=A^T`. 

maybe this should be clarified where appropriate.

jppelteret replied:
Subtle error as far as I can tell:

`[dev_P : A]_{ij} = [1/2[ \delta_{ik} \delta_{jl} + \delta_{il} \delta_{jk}] -...] A_{kl} = 1/2[A_{ij} + A_{ji}] - ...`

See the definition of the [identity_tensor](http://dealii.org/developer/doxygen/deal.II/classSymmetricTensor.html#a0d40eac7f1138a9018c41564f5712ee1) and the discussion in the intro to [step-44](http://dealii.org/developer/doxygen/deal.II/step_44.html) :-)

davydden replied:
> Subtle error

no, it's not :smile: It's just a notation/definition issue.

jppelteret replied:
I technically state this in the definition of the unit SymmetricTensor `II` that I provide in this namespace. But I've also now added some additional documentation to the definition of the projection tensor itself. Hopefully nobody will be left scratching their heads about this point!

bangerth replied:
Just a technical comment on this note above:

  Because of the symmetry of `dev_P`, it is a projection operator for symmetric 
  tensors. If `A` is not symmetric then this will return `sym(A) - 1/dim tr(A) I`.

Then it's still a *projection*, because `dev_P(dev_P(A)) = dev_P(A)` even for non-symmetric tensors.

bangerth commented:
I have this on my list. If you give me a couple more days, then I'll look over it as well.

bangerth commented on include/deal.II/physics/elasticity/standard_tensors.h:
```diff
@@ -0,0 +1,360 @@
+       * The computed result is
+       * @f[
+       *  \frac{\partial J}{\partial \mathbf{C}}
+       *   = \frac{1}{2} J \mathbf{C}^{-1}
```
Can you say how `F` and `C` are related?

jppelteret replied:
This is now stated below; Github hasn't picked up the change.

bangerth commented on include/deal.II/physics/elasticity/standard_tensors.h:
```diff
@@ -0,0 +1,360 @@
+       * @f[
+       *  \left[ \frac{\partial \mathbf{C}^{-1}}{\partial \mathbf{C}} \right]_{abcd}
+       *    := -\frac{1}{2}[ C^{-1}_{ac}C^{-1}_{bd}
+       *     + C^{-1}_{ad}C^{-1}_{bc}  ]
```
same as above

what do the indices on C represent?

jppelteret replied:
This is just the usual index notation. There is an abnormal combination of indices on the RHS, so thats why I used it. To the chosen indices {a,b,c,d} bother you?

bangerth replied:
Oh, of course. I had seen `_{ad}` and thought that that's maybe an adjugate or something, and then my brain failed to realize that `_{bc}` is clearly not something special -- got me on the wrong track and that's where I stayed.

It's fine with me. I would have chosen `ijkl`, but I don't care much one way or the other. It's going to be obvious in a displayed formula.

bangerth commented on include/deal.II/physics/physics.h:
```diff
@@ -0,0 +1,137 @@
+
+// This file only provides definition and documentation of the
+// namespace Physics. There is no necessity to include it
+// anywhere in a C++ code. Only doxygen will make use of it.
```
We have a number of headers like this, and have typically kept them in `doc/doxygen/headers`. Want to move this one there as well? This way you don't need to bother with the `#ifndef` above, etc.

jppelteret replied:
Ah, of course. I forgot about that!

bangerth commented on include/deal.II/physics/physics.h:
```diff
@@ -0,0 +1,137 @@
+   *
+   * Finally, two common tensor operators are represented by $\cdot$ and $:$
+   * operators. These respectively represent a single and double contraction over
+   * the inner tensor indices.
```
oh, I see -- thanks for providing the definition!

bangerth commented on include/deal.II/physics/physics.h:
```diff
@@ -0,0 +1,137 @@
+   *        \qquad \text{and} \qquad
+   *      \mathbf{A}^T = \overline{\mathcal{I}}:\mathbf{A} \, .
+   * @f]
+   * Note that $\mathcal{I} \neq \overline{\mathcal{I}}^T$.
```
Would it be worthwhile pointing out that in index notation, you have
```
  I_{ijkl} = \delta_{ik}\delta_{jl}
  \overline I_{ijkl} = \delta_{il}\delta_{jk}
```
with Kronecker deltas as commonly defined.

bangerth commented on include/deal.II/physics/physics.h:
```diff
@@ -0,0 +1,137 @@
+   * @f]
+   * such that
+   * @f[
+   *      \dfrac{1}{2}[\mathbf{A} + \mathbf{A}^T] = \mathcal{S}:\mathbf{A}
```
In this and the next two formulas, you basically define the right hand side in terms of the left. Maybe swap the two sides of the equality?

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+// ---------------------------------------------------------------------
```
might also go into `doc/doxygen/headers`.

jppelteret replied:
Done.

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+  * A collection of operations to assist in the transformation of tensor
+  * quantities from the reference to spatial configuration, and vice versa.
+  * These types of transformation are typically used to reexpress quantities
+  * measured or computed in configuration in terms of a second configuration.
```
-> in ONE configuration

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+  *
+  * For convenience we will predefine some commonly referenced tensors and
+  * operations.
+  * Considering the position vector $\mathbf{X}$ in the referential (material)
```
You are repeating text that you already have in the other namespace. To avoid these two texts become out of sync, I would simply say here "We will use the same notation for coordinates $x,X$ and transformations $\varphi$ as discussed for namespace ABC."

jppelteret replied:
Good idea. I'll implement this as you've suggested.

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+    /**
+     * Transformation of tensors that are defined in terms of a set of
+     * contravariant bases. Rank-1 and rank-2 contravariant tensors
+     * $\left(\bullet\right)^{\#} = \mathbf{T}$ (and its spatial counterpart
```
I don't understand the notation `\left(\bullet\right)^{\#}` -- is this common?

bangerth replied:
Also, given that you use `\flat` later on, is `\#` the same as `\sharp`? Does it render the same way?

jppelteret replied:
I wouldn't say that the notation is _common_, but its the easiest way to highlight that the tensors that these transformations are performed on are not arbitrary. This notation is from Holzapfel's book, while Wriggers (who doesn't define this collection of transformation in general, but rather only showcases a few specific examples) uses some sub- and super-script indices to denote covariant and contravariant bases. 

If I chose the latter option then I'd have to write everything in index notation and define multiple basis vectors, a convective coordinate system, ... all of which even more confusing (to me, at least).

Would you be happy if I were to make a stronger connection between this and what is presented by Holzapfel?

`\#` -> `\sharp` noted. I'll check to see if this works.

bangerth replied:
I would just say something along the following lines:
```
Following Holzapfel (2027), we will use the annotation $\left(\bullet\right)^{\#}$ to indicate that a tensor 
$\left(\bullet\right)$ is a contravariant tensor, and $\left(\bullet\right)^{\flat}$ that it is covariant. In other
words, these indices do not actually change the tensor, they just indicate the <i>kind</i> of object a
particular tensor is.
```

jppelteret replied:
I must make a note in my calender to buy his forthcoming book 😆

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+
+      /**
+      * Returns the result of the push forward transformation on a rank-2
+      * contravariant Tensor, i.e.
```
-> tensor

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+
+      /**
+      * Returns the result of the push forward transformation on a rank-2
+      * contravariant SymmetricTensor, i.e.
```
-> symmetric tensor

bangerth replied:
same everywhere below

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+    template <int dim, typename Number>
+    inline
+    Tensor<1,dim,Number>
+    transformation_contraction (const Tensor<1,dim,Number> &V,
```
I suspect that you don't want these functions to be exported to user space? In that case, just wrap these functions into an anonymous namespace:
```
namespace internal
{
  namespace Physics
  {
    namespace 
    {
      ...
    }
  }
}
```
This also has the advantage that these functions (which may become instantiated in many object files) don't all end up as exported symbols of the object files and consequently of the shared libraries -- and small object file interfaces keep linking reasonably fast.

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+    {
+      // Its significantly quicker (in 3d) to push forward
+      // each index individually
+      return contract<1,3>(F,contract<1,2>(F,contract<1,1>(F,contract<1,0>(F, H))));
```
:-)

bangerth commented on include/deal.II/physics/transformations.h:
```diff
@@ -0,0 +1,1177 @@
+          for (unsigned int K=0; K<dim; ++K)
+            for (unsigned int L=0; L<dim; ++L)
+              for (unsigned int I=0; I<dim; ++I)
+                tmp_1[i][J][K][L] += F[i][I] * H[I][J][K][L];
```
Am I right that you have to spell out all of these loops because you're working on one symmetric and one non-symmetric tensor?

jppelteret replied:
Yes, that's correct. I think that the optimal approach is for the first and the last of these loops to be manually constructed because of the various tensor symmetries. I can rewrite the middle two using the `contract<>` function since all quantities are standard tensors though.

bangerth replied:
Your choice.

bangerth commented on source/physics/elasticity/kinematics.inst.in:
```diff
@@ -0,0 +1,68 @@
+{
+    template
+    Tensor<2,deal_II_dimension,number>
+    Physics::Elasticity::Kinematics::F<deal_II_dimension,number>(const Tensor<2,deal_II_dimension,number>&);
```
I think I dimly remember that in C++98 you can only instantiate stuff in its proper namespace, not by referencing functions in a nested namespace. So you have to enter and leave the namespaces. But I may also be wrong. Let's just see what happens when this gets committed.

bangerth commented on source/physics/elasticity/standard_tensors.inst.in:
```diff
@@ -0,0 +1,41 @@
+{
+    template
+    SymmetricTensor<4,deal_II_dimension,number>
+    Physics::Elasticity::StandardTensors<deal_II_dimension>::Dev_P<number>(const Tensor<2,deal_II_dimension,number> &);
```
I think you also need to explicitly instantiate the *variables* you defined, not just the *functions*.

jppelteret commented:
@bangerth @davydden Thanks for all of the great feedback! You've given me plenty to do, and I'll get working on this and reviewing each point very shortly :-)

jppelteret commented on include/deal.II/base/tensor.h:
```diff
@@ -1921,6 +1921,47 @@ transpose (const Tensor<2,dim,Number> &t)
+ * @f[
+ *  \textrm{adj}\left(\bullet\right)
+ *   := \textrm{det}\left(\bullet\right) \; \left(\bullet\right)^{-1} \; .
+ * @f]
```
Thanks, I've now learned something new (and I now have to unlearn a falsehood): This is actually the [adjugate matrix](https://en.wikipedia.org/wiki/Adjugate_matrix) but, confusingly, is sometimes loosely referred to as the tensor adjoint (probably only by engineers though :-). I'll correct the function name and definition, and I suppose I can then forgo the addition of your note.

bangerth replied:
Ha, that's funny. Yes, that makes a lot more sense. Thanks for looking it up!

jppelteret commented on include/deal.II/physics/elasticity/kinematics.h:
```diff
@@ -0,0 +1,367 @@
+Tensor<2, dim, Number>
+Physics::Elasticity::Kinematics::F (const Tensor<2, dim, Number> &Grad_u)
+{
+// See Wriggers (2008) p 23 equ. 3.14
```
I would be very happy to do this now. I presume it entails setting up some [custom commands / alias'](http://www.stack.nl/~dimitri/doxygen/manual/custcmd.html) in `doc/doxygen/options.dox.in`?

bangerth replied:
There are two ways to define such macros. For the first one, take a look at `doc/doxygen/options.dox.in`, around line 82 and following: These commands just define macros with one argument that will simply be expanded to something else, with no further postprocessing of arguments. In your case, you want two arguments -- for this, take a look at https://www.stack.nl/~dimitri/doxygen/manual/custcmd.html .

We have a second way where we can do some actual processing of the arguments. I don't think you'll actually need this -- you just have page and equation numbers, nothing needs to happen to them -- but that can be done in a filter script that we run before doxygen even sees anything. That would be in `doc/doxygen/scripts/filter, and examples start at line 82. (What a funny coincidence that it's the same line number as above :-) )

jppelteret commented on include/deal.II/physics/elasticity/standard_tensors.h:
```diff
@@ -0,0 +1,360 @@
+       * mentioned in the class description and that returned by identity_tensor().
+       *
+       * @note Since the symmetry of the operand is assumed, this fourth-order unit
+       * tensor should only be used to operate on symmetric second-order tensors.
```
Operand -> something else: Noted.

This note was more of a word of caution that "if you apply this to a standard tensor then it doesn't behave like the 4th order identity tensor, but rather as a symmetrization operator". I could just change the note to that effect, and specifically point out that it symmetri(z)es the tensor if you perform a contraction over both of its indices.

bangerth replied:
Yes, what you just explained would be a good sentence to put into the documentation.

bangerth commented:
OK, great. Let me know once you're through. I don't think there was anything that's actually difficult to do.

bangerth commented on include/deal.II/physics/elasticity/standard_tensors.h:
```diff
@@ -0,0 +1,360 @@
+       * at the outset, seem to be a reasonable choice.
+       * However, in the author's view it makes direct implementation of the
+       * expressions for isochoric (referential) stress contributions and their
+       * linearisation simpler in practise.
```
That's not even the flag of your country!?!? ;-)

jppelteret replied:
Interestingly, as far as mother tongue language is concerned, English only ranks about 5 out of 11 for our national languages. So 🇿🇦 is definitely not representative in this case :-)

davydden commented:
w.r.t fourth order tesnors, i think to make things clear we need to introduce more than one, namely

1. `I`: `I:A=A` for all A.
2. `I^t`: `I^t:A=A^T` for all A.
3. `I^sym`: `I^sym:A=A^{sym}`
4. `I^skw`: `I^skw:A=A^{skw}`.

It's the third one which you would use, but all of them are valid per-se.

davydden commented:
> It's the third one which you would use, but all of them are valid per-se.

me and @jppelteret just discussed this issue in person (so good to be able to do it!). We agreed to stick with the current notation and not introduce anything extra. My only wish was to start with 
```
I:A = 1/2 [A^T+A]
```
as the definition of the fourth order symmetric unit tensor, from which it follows that
```
I:A=A
```
in case A is symmetric second order.

jppelteret commented on include/deal.II/physics/elasticity/standard_tensors.h:
```diff
@@ -0,0 +1,360 @@
+      static const SymmetricTensor<2, dim> I;
+
+      /**
+       * The fourth order referential/spatial unit SymmetricTensor $\mathcal{I}$.
```
After a conversation with @davydden , I now understand your concern here. I've denoted all fourth order tensors with calliagraphic font (Wriggers), but of course there's no `cal` in variable names. There's also already an `I`, so `II` had to do. I've made a note of how we distinguish vectors/rank-2 tensors/rank-4 tensors notationally the introduction to the physics namespace.  Do you find this satisfactory?

bangerth commented:
You two are the experts here. I'm just reviewing for technical clarity. Whatever you agree on, let me know when you're there and I can take another look. Although I suspect that I will just merge -- the previous version was already good, and all I had to note were text issues.

jppelteret commented on include/deal.II/physics/elasticity/kinematics.h:
```diff
@@ -0,0 +1,367 @@
+       * gradient).
+       * The result is expressed as
+       * @f[
+       *  \mathbf{d} := \frac{1}{2} \left[ \mathbf{l} + \mathbf{l}^{T} \right] \, .
```
This is now stated below; Github hasn't picked up the change

jppelteret commented on source/physics/transformations.inst.in:
```diff
@@ -0,0 +1,216 @@
+// ---------------------------------------------------------------------
+
+
+for (deal_II_dimension : DIMENSIONS; number : REAL_SCALARS)
```
@bangerth You mentioned earlier that I have to open up namespaces in the `cc` file instead of in `inst.in`. Because this instantiations file references 3 different namespaces, will I have to split it up into 3 different files?

jppelteret replied:
And if so, what would the filename convention be? Would `transformations_covariant.inst.in` etc. be acceptable?

bangerth replied:
No, you have to open and close these namespaces inside the `for` loop you have right here. Take a look at, say, `data_out_base.inst.in` in the same directory.

jppelteret commented on source/physics/elasticity/standard_tensors.inst.in:
```diff
@@ -0,0 +1,41 @@
+for (deal_II_dimension : DIMENSIONS)
+{
+    template
+    class StandardTensors<deal_II_dimension>;
```
I believe that when I instantiate a class instance, all of the static members will be instantiated as well (in this instance, `StandardTensors<dim>::I` etc. It certainly seems to work this way on my machine (clang 3.7). Is this correct in general, or should I rather instantiate each static member individually.

bangerth replied:
It *instantiates* them, but only if there is a *definition*. That's what's missing for static member variables -- these are global variables in a sense, and you've declared them, but you still need to provide a *definition* for them that can then be instantiated.

For example, if you had
```
  template <int dim> class C {
    static const int var = 13;
  };
  template class C<1>;  // explicit instantiation, but the compiler has not seen a definition for C::var
```
what you need to do is this:
```
  template <int dim> class C {
    static const int var = 13;
  };
  template <int dim> const int C<dim>::var;  // definition
  template class C<1>;  // explicit instantiation, but the compiler has not seen a definition for C::var
```

jppelteret replied:
Oh, ok we're on the same page now. I think that you might have skipped over their definitions, which are now [here](https://github.com/dealii/dealii/pull/3034/files#diff-48b3ed19fabc1d678153b1734437b782R22). So is this ok?

bangerth replied:
Yes, excellent. So any more changes that are necessary?

jppelteret commented:
I think that I've addressed all points raised. If there are any more comments or something that I've missed, I'll be happy to fix them up ASAP. If not then I'll rebase and squash this into three commits.

davydden commented on doc/doxygen/options.dox.in:
```diff
@@ -85,6 +85,10 @@ ALIASES += dealiiOperationIsMultithreaded="@note If deal.II is configured with t
 
+ALIASES += dealiiWriggersA{2}="@note For a discussion of the background of this function, see P. Wriggers: \"Nonlinear finite element methods\" (2008), and in particular formula (\2) on p. \1 (or thereabouts). "
+
+ALIASES += dealiiHolzapfelA{2}="@note For a discussion of the background of this function, see G. A. Holzapfel: \"Nonlinear solid mechanics. A Continuum Approach for Engineering\" (2007), and in particular formula (\2) on p. \1 (or thereabouts). "
```
neat!

jppelteret replied:
@bangerth's suggestion. Very neat indeed!

davydden commented on include/deal.II/physics/elasticity/standard_tensors.h:
```diff
@@ -0,0 +1,369 @@
+       *     := \dfrac{1}{2}[\{ \hat{\bullet} \} + \{ \hat{\bullet} \}^T] \, .
+       * @f]
+       *
+       * As a corollary to this, for any second-order symmetric tensor $\{ \bullet \}$
```
thanks!

jppelteret replied:
Thanks for the suggestion! I think this reads better now :-)

davydden replied:
😄

davydden commented:
i would say do rebase/squashing

bangerth commented:
Do my last two comments require any changes? Let us know once you're definitely done!

jppelteret commented:
@bangerth I think that this is good to go. I don't think that I have the willpower to look through the generated documentation again, but in any case it passes all of the tests I've written and step-44 works fine.

bangerth commented:
Good enough for me -- it looked good in the previous iteration already!

Let's give the tester time to go through everything, but from my perspective it's good to merge. Thanks for the incredible amount of work that has clearly gone into this!

jppelteret commented:
You're most welcome, and thanks in return for your patience and guidance! I can only hope that many people find this useful and add to it :-)

jppelteret commented:
Thanks as well to @davydden, especially for our animated discussion yesterday :-)

bangerth commented:
I already have a use in #3669 :-)

bangerth commented:
By the way, this patch has almost 10,000 lines. Very very impressive!

(In hindsight, this also explains why it took so long for me to go through ;-) )

jppelteret commented:
Oh wow, I didn't realise it was that big. To be fair, there are 3 versions of `step-44` in there that inflate the line count considerably! But it also explain why it took so long to polish and tidy up...

bangerth commented:
OK, it's in! Well done!

Once the documentation on the homepage is updated, want to write an email to the mailing list that announces the good news?

jppelteret commented:
Awesome! Yes, I'd be happy to do so. I'll post a draft announcement to the dev mailing list tomorrow and we can discuss its contents in the forthcoming days. I'd like to propose some guidelines on what may / may not be suitable to be added here, which I would add to the wiki at the time of the announcement.
